compiler/src/dotty/tools/dotc/typer/Typer.scala

```diff
@@ -854,8 +854,21 @@ class Typer extends Namer
     case _ => tree
   }
 
+
   def typedNamedArg(tree: untpd.NamedArg, pt: Type)(using Context): NamedArg = {
-    val arg1 = typed(tree.arg, pt)
+    /* Special case for resolving types for arguments of an annotation defined in Java.
+     * It allows that value of any type T can appear in positions where Array[T] is expected.
+     * For example, both `@Annot(5)` and `@Annot({5, 6}) are viable calls of the constructor
+     * of annotation defined as `@interface Annot { int[] value() }`
+     * We assume that calling `typedNamedArg` in context of Java implies that we are dealing
+     * with annotation contructor, as named arguments are not allowed anywhere else in Java.
+     */
+    val arg1 = pt match {
+      case AppliedType(a, typ :: Nil) if ctx.isJava && a.isRef(defn.ArrayClass) =>
+        tryAlternatively { typed(tree.arg, pt) } { typed(untpd.JavaSeqLiteral(tree.arg :: Nil, TypeTree(typ)), pt) }
```

Comment:
It's better to wrap typed tree inside a `TypedSplice`:

```suggestion
        tryAlternatively { typed(tree.arg, pt) } { 
            val elemTp = untpd.TypedSplice(TypeTree(typ))
            typed(untpd.JavaSeqLiteral(tree.arg :: Nil, elemTp), pt) 
        }
```